tools/ts-generator/types/functions.ts

```diff
@@ -266,6 +300,63 @@ const readObject = <T>(
   return obj.fromBytes(buffer, offset);
 };
 
+const extendRecords = (records: Record[], seed = 0): number => {
+  const auxBuffer = Buffer.alloc(8);
+  return records.reduce((prev, record) => {
+    let size;
+    size = writeVarintBuffer(BigInt(record.length), auxBuffer);
+    const lengthCrc = calculate(auxBuffer.slice(0, size), prev);
+    size = writeVarintBuffer(BigInt(record.attributes), auxBuffer);
+    const attrCrc = calculate(auxBuffer.slice(0, size), lengthCrc);
+    size = writeVarintBuffer(record.timestampDelta, auxBuffer);
+    const timesStampCrc = calculate(auxBuffer.slice(0, size), attrCrc);
+    size = writeVarintBuffer(BigInt(record.offsetDelta), auxBuffer);
+    const offsetDeltaCrc = calculate(auxBuffer.slice(0, size), timesStampCrc);
+    size = writeVarintBuffer(BigInt(record.keyLength), auxBuffer);
+    const keyLengthCrc = calculate(auxBuffer.slice(0, size), offsetDeltaCrc);
+    const keyCrc = calculate(record.key, keyLengthCrc);
+    size = writeVarintBuffer(BigInt(record.valueLen), auxBuffer);
+    const valueLengthCrc = calculate(auxBuffer.slice(0, size), keyCrc);
+    const valueCrc = calculate(record.value, valueLengthCrc);
+    size = writeVarintBuffer(BigInt(record.headers.length), auxBuffer);
+    return calculate(auxBuffer.slice(0, size), valueCrc);
+  }, seed);
+};
+
+const recordBatchEncode = (value: RecordBatch, buffer: IOBuf): number => {
+  let wroteBytes = 0;
+  // write header some attribute on BE format
+  const bufferHeaderCrc = Buffer.allocUnsafe(40);
+  bufferHeaderCrc.writeInt16BE(value.header.attrs, 0);
+  bufferHeaderCrc.writeInt32BE(value.header.lastOffsetDelta, 2);
+  bufferHeaderCrc.writeBigInt64BE(value.header.firstTimestamp, 6);
+  bufferHeaderCrc.writeBigInt64BE(value.header.maxTimestamp, 14);
+  bufferHeaderCrc.writeBigInt64BE(value.header.producerId, 22);
+  bufferHeaderCrc.writeInt16BE(value.header.producerEpoch, 30);
+  bufferHeaderCrc.writeInt32BE(value.header.baseSequence, 32);
+  bufferHeaderCrc.writeInt32BE(value.header.recordCount, 36);
+
+  // reserve record batch header, term, isCompressed.
+  const reserver = buffer.getReserve(70);
+  // create auxiliary iobuf
+  const iheader = new IOBuf();
+
+  wroteBytes += writeObject(iheader, RecordBatchHeader, value.header);
+  wroteBytes += writeArray(false)(value.records, buffer, (item, auxBuffer) =>
+    writeObject(auxBuffer, Record, item)
+  );
+  // Get buffer instance from iobuf
+  const headerBuffer = iheader.getIterable().slice(70);
+  // Calculate header crc, it depends on records and, some header attributes
+  const crc = extendRecords(value.records, calculate(bufferHeaderCrc));
+  headerBuffer.writeUInt32LE(crc, 17);
```

Comment:
you don't need to create a buffer to calculate the crc. You can use the 'walking crc' method to build up the hash, passing one item at a time to the caclulate method.

Reply:
Sorry, I don't understand that because, because I don't create a buffer in that line, I used iobuf reserve for writing the CRC code.  @graphcareful

Reply:
you created one for this purpose on line 342

Reply:
Umm, I used that IOBuf for operating writing buffer operation over reserve space. but I have one idea how I can remove that line `342`

Reply:
Done, I could remove that line, please check @graphcareful

Reply:
can you explain some of the 'magic numbers' here in comments, like 17, 4, 61, etc?

Reply:
Done, I put comments for that.